ge.1
.\" Copyright (c) 2022 Omar Polo <op@omarpolo.com>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.Dd September 7, 2022
.Dt GE 1
.Os
.Sh NAME
.Nm ge
.Nd export a directory over Gemini
.Sh SYNOPSIS
.Nm
.Bk -words
.Op Fl hV
.Op Fl d Ar certs-dir
.Op Fl H Ar hostname
.Op Fl p Ar port
.Op Ar directory
.Ek
.Sh DESCRIPTION
.Nm
exports the given
.Ar directory
over the Gemini protocol.
It's intended to be used interactively mostly for testing purposes,
for a full-fledged daemon look for
.Xr gmid 8 .
.Pp
The arguments are as follows:
.Bl -tag -width Ds
.It Fl d Ar certs-path
Directory where certificates are stored.
By default is
.Pa $XDG_DATA_HOME/gmid ,
i.e.\&
.Pa ~/.local/share/gmid .
.It Fl H Ar hostname
The
.Ar hostname
to use,
.Ar localhost
by default.
Certificates for the given
.Ar hostname
are searched inside the
.Ar certs-dir
specified with the
.Fl d
option.
The certificate files are named
.Ar hostname Ns .pem
and
.Ar hostname Ns .key
and are implicitly generated if not found.
.It Fl h , Fl -help
Print the usage and exit.
.It Fl p Ar port
The port to bind to, 1965 by default.
.It Fl V , Fl -version
Print the version and exit.
.It Ar directory
The root directory to serve, or the current working directory if not
specified.
.El
.Sh SEE ALSO
.Xr gmid 8
.Sh ACKNOWLEDGEMENTS
.Nm
uses the
.Dq Flexible and Economical
UTF-8 decoder written by
.An Bjoern Hoehrmann .
.Sh AUTHORS
.An -nosplit
The
.Nm
program was written by
.An Omar Polo Aq Mt op@omarpolo.com .


ChangeLog
2022-09-10  Omar Polo  <op@omarpolo.com>

	* parse.y (string): retire the deprecated `mime' and `map' config options

2022-09-08  Omar Polo  <op@omarpolo.com>

	* gmid.c (main): change the flags to be consisent with other OpenBSD daemons

2022-09-07  Omar Polo  <op@omarpolo.com>

	* ge.c (main): add `ge': small program to quickly export a directory over Gemini.

2022-09-06  Omar Polo  <op@omarpolo.com>

	* server.c: drop CGI support.
	* sandbox.c: optionally disable the sandbox on non-OpenBSD when fastcgi or proxy are used.  New sockets can't be opened under FreeBSD' capsicum(4) and a huge pain in the arse under Linux' seccomp/landlock.  Patches welcome!

2022-07-07  Omar Polo  <op@omarpolo.com>

	* log.c (logger_main): fix timestamps in log files.  Reported by Karl Jeacle, thanks!

	* dirs.c (scandir_fd): drop d_reclen; it's not available on DragonflyBSD (at least.)

2022-07-07  Anna “CyberTailor”

	* Makefile: include contrib/ in dist

2022-07-04  Omar Polo  <op@omarpolo.com>

	* contrib/gmid: remove OpenBSD' rc file: it's now maintained in the ports tree

	* configure (VERSION): 1.8.4 tagged

	* server.c (check_path): log when it fails to open a file because of permissions
	(client_write): encode file names in the directory index, spotted by cage
	(client_write): add a trailing / for dirs in the directory index.

	* iri.c (parse_path): allow '@' and ':' in paths

2022-04-08  Omar Polo  <op@omarpolo.com>

	* mime.c (mime): do a binary search to match the MIME time

	* gmid.c (listener_main): don't load the default mime listing when
	`types' is used.  (text/gemini -> gmi/gemini is still hardcoded)

	* mime.c (add_mime): fix memory leak in the mime handling

2022-04-07  Omar Polo  <op@omarpolo.com>

	* gmid.conf.5: move config file documentation into a new man page

	* parse.y (option): deprecate `map' rule in favour of the `types' block

2022-03-27  Omar Polo  <op@omarpolo.com>

	* configure (VERSION): 1.8.3 tagged

	* server.c (start_cgi): fix a possible out-of-bound access in the
	CGI handling.

2022-03-26  Omar Polo  <op@omarpolo.com>

	* configure (VERSION): 1.8.2 tagged

	* server.c (handle_imsg_cgi_res): fix a CGI timing issue: if a
	connection handled by a CGI scripts is interrupted with the right
	timing it causes the server process to exit with "fatal in
	client_by_id: invalid id X"

2022-02-26  Omar Polo  <op@omarpolo.com>

	* parse.y: add the types block

2022-02-13  Omar Polo  <op@omarpolo.com>

	* sandbox.c (filter): tightens seccomp policy wrt openat: allow
	only with the O_RDONLY flag.

2022-02-13  Tobias Berger  <tobi.berger13@gmail.com>

	* sandbox.c (filter): allow fstatat64, llseek and sigreturn;
	needed by glibc on armv7

2022-02-10  Omar Polo  <op@omarpolo.com>

	* configure (VERSION): 1.8.1 tagged

	* sandbox.c (open_landlock): fix landlock usage

2022-01-30  Omar Polo  <op@omarpolo.com>

	* configure (VERSION): 1.8 “Lightbulb Sun” tagged

	* proxy.c (proxy_setup_tls): allow to specify a custom hostname as SNI for the TLS handshake with the proxied host.

2022-01-02  Omar Polo  <op@omarpolo.com>

	* server.c (matched_proxy): allow to specify multiple proxy blocks and add matching rules

	* parse.y (servbody): relax the strict ordering of options, location and proxy blocks

2022-01-01  Omar Polo  <op@omarpolo.com>

	* proxy.c (proxy_init): add support for client certificate when proxying

2021-12-29  Omar Polo  <op@omarpolo.com>

	* proxy.c (proxy_init): add proxying support via `proxy relay-to'

2021-12-11  Max  <vdrummer@posteo.net>

	* sandbox.c (filter): [seccomp]: allow ugetrlimit(2), needed by glibc on armv7l

2021-12-09  Omar Polo  <op@omarpolo.com>

	* server.c (client_read): don't check if the port in the request is the same we're listening on.  Suggested by Allen Sobot, thanks!
	* configure: add --prefix=... long argument.

2021-11-16  Omar Polo  <op@omarpolo.com>

	* configure (singletest): fix the cross-compilation: don't run the test binaries just built.  Suggested by Nikolay Korotkiy (@sikmir), thanks!

2021-10-24  Omar Polo  <op@omarpolo.com>

	* log.c (log_request): fix "double slash" in logs: gmid ended up printing two slashes between the hostname and the path when logging the request IRI.

2021-10-15  Stephen Gregoratto  <dev@sgregoratto.me>

	* gmid.c (add_keypair): implement OCSP stapling support

2021-10-13  Omar Polo  <op@omarpolo.com>

	* server.c (do_accept): don't die on ECONNABORTED (``Software caused connection abort'')

2021-10-11  Omar Polo  <op@omarpolo.com>

	* contrib/renew-certs: add script to automatically renew self-signed certificates

2021-10-09  Omar Polo  <op@omarpolo.com>

	* parse.y (print_conf): multiple -n to dump the parsed configuration

	* contrib/gencert: add gencert, a simple script to generate self-signed certs

2021-10-04  Omar Polo  <op@omarpolo.com>

	* regress/lib.sh (raw): reduced the timeout time for single checks from 30 to 10 seconds

	* regress/runtime: regression test restructured.  still implemented as a set of POSIX-scripts

2021-10-02  Omar Polo  <op@omarpolo.com>

	* server.c (client_read): reworked the internal structure.  Now we leverage libevent' bufferevents more.
	(cgi_read): be more strict with CGI scripts: don't pass through illegal CGI responses.

2021-09-26  Omar Polo  <op@omarpolo.com>

	* fcgi.c (fcgi_req): the FastCGI implementation is fully asynchronous

2021-09-24  Omar Polo  <op@omarpolo.com>

	* configure (VERSION): 1.7.4 tagged

	* server.c (check_for_cgi): fix out-of-bound access, found with lots of help from cage, thanks!

2021-09-19  Omar Polo  <op@omarpolo.com>

	* sandbox.c (gmid_create_landlock_rs): [linux] use landlock on the server and logger process to reduce the fs access

	* configure (VERSION): 1.7.3 tagged

2021-09-17  Omar Polo  <op@omarpolo.com>

	* log.c (logger_main): fix syslog logging on FreeBSD

2021-08-24  Omar Polo  <op@omarpolo.com>

	* gmid.c (main): don't crash on -n without -c

2021-07-29  Anna “CyberTailor”


ge.c
/*
 * Copyright (c) 2022 Omar Polo <op@omarpolo.com>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include "gmid.h"

#include <sys/stat.h>
#include <sys/wait.h>

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <libgen.h>
#include <signal.h>
#include <string.h>
#include <unistd.h>

struct imsgbuf ibuf, logibuf;
struct conf conf;

struct fcgi fcgi[FCGI_MAX];	/* just because it's referenced */
struct vhosthead hosts;


static const struct option opts[] = {
	{"help",	no_argument,	NULL,	'h'},
	{"version",	no_argument,	NULL,	'V'},
	{NULL,		0,		NULL,	0},
};

void
load_local_cert(struct vhost *h, const char *hostname, const char *dir)
{
	char *cert, *key;

	if (asprintf(&cert, "%s/%s.cert.pem", dir, hostname) == -1)
		errx(1, "asprintf");
	if (asprintf(&key, "%s/%s.key.pem", dir, hostname) == -1)
		errx(1, "asprintf");

	if (access(cert, R_OK) == -1 || access(key, R_OK) == -1)
		gen_certificate(hostname, cert, key);

	h->cert = cert;
	h->key = key;
	h->domain = hostname;
}

/* wrapper around dirname(3).  dn must be PATH_MAX+1 at least. */
static void
pdirname(const char *path, char *dn)
{
	char	 p[PATH_MAX+1];
	char	*t;

	strlcpy(p, path, sizeof(p));
	t = dirname(p);
	memmove(dn, t, strlen(t)+1);
}

static void
mkdirs(const char *path, mode_t mode)
{
	char	dname[PATH_MAX+1];

	pdirname(path, dname);
	if (!strcmp(dname, "/"))
		return;
	mkdirs(dname, mode);
	if (mkdir(path, mode) != 0 && errno != EEXIST)
		fatal("can't mkdir %s: %s", path, strerror(errno));
}

/* $XDG_DATA_HOME/gmid */
char *
data_dir(void)
{
	const char *home, *xdg;
	char *t;

	if ((xdg = getenv("XDG_DATA_HOME")) == NULL) {
		if ((home = getenv("HOME")) == NULL)
			errx(1, "XDG_DATA_HOME and HOME both empty");
		if (asprintf(&t, "%s/.local/share/gmid", home) == -1)
			err(1, "asprintf");
	} else {
		if (asprintf(&t, "%s/gmid", xdg) == -1)
			err(1, "asprintf");
	}

	mkdirs(t, 0755);
	return t;
}

static void
logger_init(void)
{
	int p[2];

	if (socketpair(AF_UNIX, SOCK_STREAM, PF_UNSPEC, p) == -1)
		err(1, "socketpair");

	switch (fork()) {
	case -1:
		err(1, "fork");
	case 0:
		close(p[0]);
		setproctitle("logger");
		imsg_init(&logibuf, p[1]);
		_exit(logger_main(p[1], &logibuf));
	default:
		close(p[1]);
		imsg_init(&logibuf, p[0]);
		return;
	}
}

static int
serve(const char *host, int port, const char *dir, struct tls *ctx)
{
	struct addrinfo hints, *res, *res0;
	int error, saved_errno, sock = -1;
	const char *cause = NULL;
	char service[32];

	if (snprintf(service, sizeof(service), "%d", port) < 0)
		fatal("snprintf");

	memset(&hints, 0, sizeof(hints));
	hints.ai_family = AF_UNSPEC;
	hints.ai_socktype = SOCK_STREAM;
	hints.ai_flags = AI_PASSIVE;
	error = getaddrinfo(host, service, &hints, &res0);
	if (error)
		fatal("%s", gai_strerror(error));
	for (res = res0; res; res = res->ai_next) {
		sock = socket(res->ai_family, res->ai_socktype,
		    res->ai_protocol);
		if (sock == -1) {
			cause = "socket";
			continue;
		}

		if (bind(sock, res->ai_addr, res->ai_addrlen) == -1) {
			cause = "bind";
			saved_errno = errno;
			close(sock);
			errno = saved_errno;
			continue;
		}

		if (listen(sock, 5) == -1)
			fatal("listen");

		/*
		 * for the time being, we're happy as soon as
		 * something binds.
		 */
		break;
	}

	if (sock == -1)
		fatal("%s", cause);
	freeaddrinfo(res0);

	log_notice(NULL, "serving %s on port %d", dir, port);
	loop(ctx, sock, -1, NULL);
	return 0;
}

static __dead void
usage(void)
{
	fprintf(stderr,
	    "Version: " GE_STRING "\n"
	    "Usage: %s [-hVv] [-d certs-dir] [-H hostname] [-p port] [dir]\n",
	    getprogname());
	exit(1);
}

int
main(int argc, char **argv)
{
	struct tls_config *tlsconf;
	struct tls *ctx;
	struct vhost *host;
	struct location *loc;
	const char *errstr, *certs_dir = NULL, *hostname = "localhost";
	char path[PATH_MAX];
	int ch;

	logger_init();
	conf.port = 1965;

	while ((ch = getopt_long(argc, argv, "d:H:hp:Vv", opts, NULL)) != -1) {
		switch (ch) {
		case 'd':
			certs_dir = optarg;
			break;
		case 'H':
			hostname = optarg;
			break;
		case 'h':
			usage();
			break;
		case 'p':
			conf.port = strtonum(optarg, 0, UINT16_MAX, &errstr);
			if (errstr)
				fatal("port number is %s: %s", errstr, optarg);
			break;
		case 'V':
			puts("Version: " GE_STRING);
			return 0;
		default:
			usage();
			break;
		}
	}
	argc -= optind;
	argv += optind;

	if (argc > 1)
		usage();

	/* prepare the configuration */
	conf.verbose = 1;
	init_mime(&conf.mime);

	if (certs_dir == NULL)
		certs_dir = data_dir();

	if (load_default_mime(&conf.mime) == -1)
		fatal("can't load default mime types");
	sort_mime(&conf.mime);

	/* set up the implicit vhost and location */

	host = xcalloc(1, sizeof(*host));
	TAILQ_INSERT_HEAD(&hosts, host, vhosts);

	loc = xcalloc(1, sizeof(*loc));
	loc->fcgi = -1;
	TAILQ_INSERT_HEAD(&host->locations, loc, locations);

	load_local_cert(host, hostname, certs_dir);

	host->domain = "*";
	loc->auto_index = 1;
	loc->match = "*";

	if (*argv == NULL) {
		if (getcwd(path, sizeof(path)) == NULL)
			fatal("getcwd");
		loc->dir = path;
	} else
		loc->dir = absolutify_path(*argv);

	if ((loc->dirfd = open(loc->dir, O_RDONLY|O_DIRECTORY)) == -1)
		fatal("can't open %s", loc->dir);

	/* setup tls */

	if ((tlsconf = tls_config_new()) == NULL)
		fatal("tls_config_new"); /* XXX: fatalx */

	/* optionally accept client certs but don't try to verify them */
	tls_config_verify_client_optional(tlsconf);
	tls_config_insecure_noverifycert(tlsconf);

	if ((ctx = tls_server()) == NULL)
		fatal("tls_server failure"); /* XXX: fatalx */

	if (tls_config_set_keypair_file(tlsconf, host->cert, host->key))
		fatal("can't load the keypair (%s, %s)",
		    host->cert, host->key);

	if (tls_configure(ctx, tlsconf) == -1)
		fatal("tls_configure: %s", tls_error(ctx));

	/* start the server */
	signal(SIGPIPE, SIG_IGN);
	setproctitle("%s", loc->dir);
	return serve(hostname, conf.port, loc->dir, ctx);
}
